### PR TITLE
feat: try resolve conflict of delete and compact

### LINKED ISSUE #11729: Feature:  distributed execution of `delete` statement
**Summary**

- distributed execution of the `delete` statement
- refine the logic of  conflict resolution 
  avoid aborting deletion if possible (e.g. if compaction happens to the newly happened data which are not touched during deletion)

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. detect resolvable conflict between deletion and other mutations, merge snapshots of them. 
2. add an enum `ConflictResolveContext` for resolving other kinds of conflict in the future. 

Closes #11729

### CODE REVIEW

**Comment by SkyFan2002** [comment]
PTAL @dantengsky @zhyass

**Comment by dantengsky** [comment]
great :+1:  

we need unit test to cover the new conflict resolving logics

File: src/query/storages/fuse/src/operations/common/mutation_accumulator.rs [hunk 1]
```diff
@@ -178,6 +193,13 @@ impl MutationAccumulator {
 
         let chunk_size = self.ctx.get_settings().get_max_storage_io_requests()? as usize;
         let segment_indices = self.mutations.keys().cloned().collect::<Vec<_>>();
+        let removed_segments = segment_indices
+            .iter()
+            .map(|i| self.base_segments[*i].clone())
+            .collect::<Vec<_>>();
+        let mut added_segments = vec![];
+        let mut removed_statistics = Statistics::default();
+        let mut added_statistics = Statistics::default();
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/common/mutation_accumulator.rs:202]: is it feasible to have this vars only if `self.kind` is `MutationKind::Delete`?

File: src/query/storages/fuse/src/operations/common/snapshot_generator.rs [hunk 2]
```diff
@@ -59,11 +61,32 @@ pub trait SnapshotGenerator {
...
+#[derive(Clone, serde::Serialize, serde::Deserialize, Debug, PartialEq, Default)]
+pub enum ConflictResolveContext {
+    Delete(Box<DeleteConflictResolveContext>),
+    Compact,
+    Update,
+    Replace,
+    Recluster,
+    Insert,
+    #[default]
+    Uninitialized,
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/common/snapshot_generator.rs:81]: is it possible to assign a non-`Uninitialized` resolve context to `MutationGenerator`? so that some relevant `unreachable` match branches seem to be avoidable?  

or postponing the refactoring to another dedicated PR is also fine.

Comment (SkyFan2002) [reply src/query/storages/fuse/src/operations/common/snapshot_generator.rs:81]: Good idea, I will refactor in another PR

**Comment by dantengsky** [comment]
LGTM

@BohuTANG I think this PR is a good candidate, and ready for the "long run" test, shall we hold off merging this PR for a little while, till the "long run" test passed?

**Comment by BohuTANG** [comment]
> LGTM
> 
> @BohuTANG I think this PR is a good candidate, and ready for the "long run" test, shall we hold off merging this PR for a little while, till the "long run" test passed?

Longrun will ready soon.

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):


```

**Comment by BohuTANG** [comment]
askbend:summary

**Comment by BohuTANG** [comment]
askbend:summary

**Comment by BohuTANG** [comment]
## PR Summary(By [llmchain.rs](https://github.com/shafishlabs/llmchain.rs)):
* **Interpreter and Delete Changes**
- Removed unnecessary code from interpreter.rs and interpreter_delete.rs
- Removed GlobalIORuntime and TableLockHandlerWrapper imports
- Removed table lock heartbeat and refresh table code
- Updated execute2() function to be awaited

* **Pipeline Builder and Conflict Tests**
- Added MutationKind import to pipeline_builder.rs
- Added new argument to build_row_fetcher_pipeline function call
- Added conflict resolution tests for Fuse storage in conflict.rs
- Added conflict module to Fuse storage integration tests

* **Fuse Storage Changes**
- Added MutationKind enum to common module
- FuseTable struct now takes a MutationKind parameter in create_append_processor method
- Conflict enum now has a ResolvableDelete variant
- Added MutationKind and ConflictResolveContext enums to mutation_accumulator.rs
- MutationAccumulator struct now has a kind field of type MutationKind
- CommitMeta struct now has a conflict_resolve_context field of type ConflictResolveContext
- Added conflict_resolve_context field to CommitSink struct in sink_commit.rs
- Added mutation_kind field to TableMutationAggregator struct in transform_mutation_aggregator.rs
- Added ConflictResolveContext enum with variants to snapshot_generator.rs
- Added detect_conflicts method to MutationGenerator in snapshot_generator.rs
- Added ctx field of type ConflictResolveContext to MutationGenerator in snapshot_generator.rs
- Modified fill_snapshot method in MutationGenerator to handle conflicts and ResolvableDelete variant in snapshot_generator.rs
- Added MutationKind enum with variants to snapshot_generator.rs
- Modified CompactAggregator struct and FuseTable struct to use new enums in their constructors

* **Replace and Update Changes**
- Added MutationKind import to replace.rs and update.rs
- Modified chain_mutation_pipes function call to include MutationKind parameter

* **Delete Test Deletion**
- Deleted unnecessary test file 03_0025_delete_from in sqllogictests/suites/base/

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 here is the summary of major changes in commit:

* Clarified the context for detecting conflicts during mutation operations
    - Added MutationKind enum, to indicate the type of mutation during commit.
    - Encapsulated the context for resolving conflicts into ConflictResolveContext. 
    - MutationAccumulator passes the context when generating new snapshot.
    - SnapshotGenerator uses this context to resolve conflicts in a specific way, for different mutation types.

* Added tests for resolving delete conflicts 
    - Tests both unresolvable and resolvable delete conflicts.

* Removed unused code
    - Removed unused imports 
    - Removed logics related to table locks, as they are handled elsewhere now.

* **Clarified conflict resolution logic**
    - Refactored conflict resolution to encapsulate context 

* **Isolated delete conflict resolution**  
    - Handle delete conflicts differently by using a DeleteConflictResolveContext

* **Extracted common logic**
   - Extracted common functions related to statistics merging

So in summary, the major goal is to clarify and isolate the conflict resolution logic, by passing proper context and encapsulating different contexts for different mutations.

**Comment by dantengsky** [comment]
closing this PR, since PR #11894 contains (and has enhanced) this PR.

@SkyFan2002  If I misunderstand, please feel free to re-open this PR
